osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,196 @@
+import asyncio
+import datetime
+import furl
+import logging
+
+import aiohttp
+import bcrypt
+from PyQt5.QtWidgets import QMessageBox
+
+from osfoffline import settings
+from osfoffline.database_manager.db import session
+from osfoffline.database_manager.models import User
+from osfoffline.database_manager.utils import save
+from osfoffline.polling_osf_manager.remote_objects import RemoteUser
+
+
+def generate_hash(password):
+    """ Generates a password hash using `bcrypt`.
+        Number of rounds for salt generation is 12.
+
+    :return: hashed password
+    """
+    return bcrypt.hashpw(
+        password.encode('utf-8'),
+        bcrypt.gensalt(12)
+    )
+
+token_request_body = json.loads(
+    '{"data": {"type": "tokens", "attributes": {"name": "OSF-Offline {}", "scopes": "osf.full_write"}}}'.format(datetime.date.today())
+)
+
+
+class AuthClient(object):
+    """Manages authorization flow """
+    def __init__(self):
+        self.API_URL = furl.furl(settings.API_BASE)
+        self.failed_login = False
+
+    @asyncio.coroutine
+    def _authenticate(self, username, password):
+        """ Tries to use standard auth to authenticate and create a personal access token through APIv2
+
+            :return: personal_access_token or None
+        """
+        token_url = self.API_URL.path.add('/v2/tokens/create/')
+        token_url.args['username'] = username
+        token_url.args['password'] = password
```

Comment:
https://github.com/CenterForOpenScience/osf.io/blob/develop/framework/auth/cas.py#L66

AFAIK, this is the only way to avoid a terribly messy situation involving pretending to be a web browser to CAS, faking a csrf-token, and ultimately using a cookie to make a request similar to this, only with the password in the request body. You can ask Michael about it if you want more details, but this approach is preferable.